harness/api_checkers.go

```diff
@@ -0,0 +1,400 @@
+// Package harness provides a suite of API compatibility checks. They were originally
+// ported from the OpenTracing Python library's "harness" module.
+package harness
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+// APICheckCapabilities describes options used by APICheckSuite when testing a Tracer.
+type APICheckCapabilities struct {
+	CheckBaggageValues bool          // whether to check for propagation of baggage values
+	CheckExtract       bool          // whether to check if extracting contexts from carriers works
+	CheckInject        bool          // whether to check if injecting contexts works
+	Probe              APICheckProbe // optional interface providing methods to check recorded data
+}
+
+// APICheckProbe exposes methods for testing data recorded by a Tracer.
+type APICheckProbe interface {
+	SameTrace(first, second opentracing.Span) bool // whether two spans are from the same trace
+	SameTraceContext(opentracing.Span, opentracing.SpanContext) bool
+}
+
+// APICheckSuite is a testify suite for checking a Tracer against the OpenTracing API.
+type APICheckSuite struct {
+	suite.Suite
+	opts      APICheckCapabilities
+	newTracer func() (tracer opentracing.Tracer, closer func())
+	tracer    opentracing.Tracer
+	closer    func()
+}
+
+// BeforeTest creates a tracer for this specific test invocation.
+func (s *APICheckSuite) BeforeTest(suiteName, testName string) {
+	s.tracer, s.closer = s.newTracer()
+	if s.tracer == nil {
+		panic(fmt.Sprintf("newTracer returned nil Tracer before running %s, %s", suiteName, testName))
+	}
+}
+
+// AfterTest closes the tracer, and clears the test-specific tracer.
+func (s *APICheckSuite) AfterTest(suiteName, testName string) {
+	if s.closer != nil {
+		s.closer()
+	}
+	s.tracer, s.closer = nil, nil
+}
+
+// NewAPICheckSuite returns a testify suite for checking a Tracer against the OpenTracing API.
+// It is provided a function that will be executed to create and destroy a tracer for each test
+// in the suite, and API test options described by APICheckCapabilities.
+func NewAPICheckSuite(
+	newTracer func() (tracer opentracing.Tracer, closer func()),
+	opts APICheckCapabilities) *APICheckSuite {
```

Comment:
please put the argument `opts APICheckCapabilities` on it's own line, it's hard to read in this form.